keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,274 @@
+
+As a result metric data will not be invalid, just data will be more coarse when the metric type is a counter. In the case of a gauge type metric the data will be invalid but the number of series would be bound. One exception to our treatment of labels will be histograms (specifically the label which denotes buckets). If we explicitly declare a whitelist of acceptable values for histogram buckets, then we will simply omit buckets which are not in that list, since we can preserve data fidelity for histogram buckets by bucket omission (for everything except the catch-all bucket).
+
+Following is an example used to demonstrate the proposed solution.
+
+Let's say we have some client code somewhere which instantiates our metric from above:
+
+```golang
+	SomeMetric = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "some_metric",
+			Help:           "alalalala",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"label_too_many_values"},
+	)
+```
+
+And some other code which increments our metric:
+
+```golang
+	var label_value string
+	label_value = "1"
+	SomeMetric.WithLabelValues(label_value).Inc()
+	label_value = "2"
+	SomeMetric.WithLabelValues(label_value).Inc()
+	label_value = "3"
+	SomeMetric.WithLabelValues(label_value).Inc()
+```
+
+In this case we would expect our raw prometheus metric output to look like this:
+
+```prometheus
+# HELP some_metric alalalala
+# TYPE some_metric counter
+some_metric{label_too_many_values="1"} 1
+some_metric{label_too_many_values="2"} 1
+some_metric{label_too_many_values="3"} 1
+```
+This would not change. What would change is if we encounter label values outside our explicit allowlist of values. So if where to encounter this malicious piece of code:
+
+
+```golang
+	var label_value string
+	for i=0;i<1000000;i++ {
+		label_value = strconv.Itoa(i+4) // bring us out of our expected range
+		SomeMetric.WithLabelValues(label_value).Inc()
+	}
+```
+
+Then in existing Kubernetes components, we would have a terrible memory leak, since we would effectively create a million metrics (one per unique label value). If we curled our metrics endpoint it would thus look something like this:
+
+```prometheus
+# HELP some_metric alalalala
+# TYPE some_metric counter
+some_metric{label_too_many_values="1"} 1
+some_metric{label_too_many_values="2"} 1
+some_metric{label_too_many_values="3"} 1
+some_metric{label_too_many_values="4"} 1
+... //
+... //
+... // zillion metrics here
+some_metric{label_too_many_values="1000003"} 1
+```
+
+In total we would have more than a million of series for one metric, which is known to cause a memory problem.
+
+With our cardinality enforcer in place, we would expect the output to look like this:
+
+```prometheus
+# HELP some_metric alalalala
+# TYPE some_metric counter
+some_metric{label_too_many_values="1"} 1
+some_metric{label_too_many_values="2"} 1
+some_metric{label_too_many_values="3"} 1
+some_metric{label_too_many_values="unexpected"} 1000000
```

Comment:
nit: Should we blacklist this label value across the code base?  Or are we ok with collisions with the unexpected label.

Reply:
You mean across metrics? My original intent was to target whitelists for a metric label (so the conjunction of '__metric__' and 'label_too_many_values'.

Reply:
I mean, should we disallow calling `WithLabelValues("unexpected")`?  It doesn't return an error today, so i'm not entirely sure how we would do it.  Maybe we should log a warning when someone does that?

Reply:
Won't the default behavior just be the right one?

Reply:
The default behavior would be to lump together label values that are the string literal "unexpected" with label values that were filtered out with the whitelist.  Given that "unexpected" is now a special cased label, i'm suggesting we could/should discourage its use as a normal label value.

Reply:
You'd have to explicitly whitelist a label value called 'unexpected' for that to happen.

Reply:
Agreed.  It probably wouldn't ever happen.  Someone would have to use the "unexpected" value in a metric label, and then want to whitelist values for that label, including the literal "unexpected"...